sway/desktop/render.c

```diff
@@ -70,6 +72,33 @@ static void scissor_output(struct wlr_output *wlr_output,
 	wlr_renderer_scissor(renderer, &box);
 }
 
+static void set_scale_filter(struct wlr_output *wlr_output,
+		struct wlr_texture *texture, enum scale_filter_mode scale_filter) {
+	if (!wlr_texture_is_gles2(texture)) {
+		return;
+	}
+
+	struct wlr_gles2_texture_attribs attribs;
+	wlr_gles2_texture_get_attribs(texture, &attribs);
+
+	switch (scale_filter) {
+	case SCALE_FILTER_LINEAR:
+		glTexParameteri(attribs.target, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+		break;
+	case SCALE_FILTER_NEAREST:
+		glTexParameteri(attribs.target, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+		break;
+	case SCALE_FILTER_SMART:
+		if (ceilf(wlr_output->scale) == wlr_output->scale) {
+			glTexParameteri(attribs.target, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+		}
+		break;
+	default:
+		return;
```

Comment:
We should probably add an assertion here

Reply:
Hmm, this actually is reachable, although I didn't expect it.

When sway starts it gets to render with scale_filter still at SCALE_FILTER_NONE. I intended NONE as a sentinal value for unset displays so output config merging is correct.